src/deploy.ts

```diff
@@ -1,209 +1,378 @@
+      [config.token],
+    );
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(config.mailbox, config.interchainGasPaymaster),
+    );
+    return router;
+  }
+
+  protected async deployNative(
+    chain: ChainName,
+    config: HypNativeConfig,
+  ): Promise<HypNative> {
+    const router = await this.deployContractFromFactory(
+      chain,
+      new HypNative__factory(),
+      'HypNative',
+      [],
+    );
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(config.mailbox, config.interchainGasPaymaster),
+    );
+    return router;
+  }
+
+  protected async deploySynthetic(
+    chain: ChainName,
+    config: HypERC20Config,
+  ): Promise<HypERC20> {
+    const router = await this.deployContractFromFactory(
+      chain,
+      new HypERC20__factory(),
+      'HypERC20',
+      [config.decimals],
+    );
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(
+        config.mailbox,
+        config.interchainGasPaymaster,
+        config.totalSupply,
+        config.name,
+        config.symbol,
       ),
-      factories,
-      options,
     );
+    return router;
   }
 
   router(contracts: HyperlaneContracts<HypERC20Factories>) {
     return contracts.router;
   }
 
   async deployContracts(chain: ChainName, config: HypERC20Config) {
+    let router: HypERC20 | HypERC20Collateral | HypNative;
     if (isCollateralConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypERC20Collateral__factory(),
-        'HypERC20Collateral',
-        [config.token],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(config.mailbox, config.interchainGasPaymaster),
-      );
-      return { router };
-    } else if (isSyntheticConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypERC20__factory(),
-        'HypERC20',
-        [],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(
-          config.mailbox,
-          config.interchainGasPaymaster,
-          config.totalSupply,
-          config.name,
-          config.symbol,
-        ),
-      );
-      return { router };
+      router = await this.deployCollateral(chain, config);
     } else if (isNativeConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypNative__factory(),
-        'HypNative',
-        [],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(config.mailbox, config.interchainGasPaymaster),
-      );
-      return { router };
+      router = await this.deployNative(chain, config);
+    } else if (isSyntheticConfig(config)) {
+      router = await this.deploySynthetic(chain, config);
+    } else {
+      throw new Error('Invalid ERC20 token router config');
+    }
+    return { router };
```

Comment:
This is probably outside the scope of this PR, but it would be nice to have the keys here reflect the type of token so that we can construct the right contract wrappers. Also helps with the artifacts indicating to users which is which

Reply:
This is related to `Router output is generic, not which ones is native vs which one is synthetic` on https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1863 cc @jmrossy

Reply:
see https://github.com/hyperlane-xyz/hyperlane-token/pull/61/